lib/navigation/preprocess_encoders.cpp

```diff
@@ -16,41 +31,89 @@ core::EncoderData EncodersPreprocessor::processData(const core::EncoderData)
+}
+
+template<std::size_t N>
+Quartile EncodersPreprocessor::getQuartiles(std::array<std::uint32_t, N> &reliable_data){
+  std::sort(reliable_data.begin(), reliable_data.end());
+  Quartile quartile;
+  quartile.q1 = getQuartile(reliable_data,0.25);
+  quartile.median = getQuartile(reliable_data, 0.5);
+  quartile.q3 = getQuartile(reliable_data, 0.75);
+  return quartile;
+}
+
+
 core::EncoderData EncodersPreprocessor::detectOutliers(const core::EncoderData encoder_data)
 {
-  /*
-  TODOLater: implement
-  rough process:
-  - get q1, median, q3 of encoder array
-  - define upper & lower bounds as (-)1.5*inter-quatrile range
-  - if any datapoint outwith this, set
-  num_outliers_per_encoder_[i] += 1
-  -set outlier points as median
-
-  -also has to be able to handle 1 unreliable sensor
-  -also figure out return type/ what we update and update
-  documentation as appropriate
-  */
-  return {0, 0, 0, 0};
+  core::EncoderData encoder_data_copy;
+  std::copy(encoder_data.begin(), encoder_data.end(), encoder_data_copy.begin());
+  Quartile quartiles;
+  if (num_reliable_encoders_ == core::kNumEncoders - 1) {
+    std::array<uint32_t, core::kNumEncoders - 1> reliable_data;
+    std::size_t counter = 0;
+    for (std::size_t i = 0; i < encoder_data.size(); ++i) {
+      if (encoders_reliable.at(i)) {
+        reliable_data.at(counter) = encoder_data.at(i);
+        ++counter;
+      }
+    }
+    quartiles = getQuartiles(reliable_data);
+  } else {
+    core::EncoderData reliable_data;
+    std::copy(
+      encoder_data.begin(), encoder_data.end(), reliable_data.begin());
+    quartiles = getQuartiles(reliable_data); 
+  }
+  const core::Float iqr         = quartiles.q3 - quartiles.q1;
+  const core::Float upper_bound = quartiles.median + 1.5 * iqr;
+  const core::Float lower_bound = quartiles.median - 1.5 * iqr;
+
+  for (std::size_t i = 0; i < encoder_data_copy.size(); ++i) {
+    if (encoder_data_copy.at(i) > upper_bound || encoder_data_copy.at(i) < lower_bound || encoders_reliable.at(i) == false) {
+      encoder_data_copy.at(i)         = quartiles.median;
+
+      if(encoders_reliable.at(i)){
+        encoder_outliers.at(i) = encoder_outliers.at(i) + 1;
+      }
+    } else {
+      encoder_outliers.at(i) = 0;
+    }
+  }
+
+  return encoder_data_copy;
 }
 
-void EncodersPreprocessor::checkReliable(const core::EncoderData &encoder_data)
+
+
+SensorChecks EncodersPreprocessor::checkReliable()
 {
-  /*
-  TODOLater: implement
-  rough process:
-  - check how many times an individual encoder
-  has been an outlier in a row (outlier encoders)
-  -if num_outliers_per_encoder_[i] > n (tbd), mark encoder as unreliable
-  in reliable encoders.
-  -if number unreliable in reliable_encoders > 1, fail state
-
-  -also figure out return type/ what we update and update
-  documentation as appropriate
-  */
+  for (int i = 0; i < core::kNumEncoders; ++i) {
+    if (encoder_outliers.at(i) > max_consecutives) {  // for now assuming n to be 10
```

Comment:
maybe rename `max_consecutives` for clarity

Reply:
Also `max_consecutives` is a class member so needs an underscore after it (just a naming convention).